### PR TITLE
PSP-4709 : Project Search and List view (Manage projects)

### PR DESCRIPTION
**Pull request by eddherrera**

### CODE REVIEW

File: source/backend/api/Areas/Projects/Controllers/ProjectController.cs [hunk 1]
```diff
@@ -22,12 +22,8 @@ namespace Pims.Api.Areas.Projects.Controllers
     [Route("[area]")]
     public class ProjectController : ControllerBase
     {
-        #region Variables
         private readonly IProjectService _projectService;
         private readonly IMapper _mapper;
-        #endregion
-
-        #region Constructors
```
Comment (FuriousLlama) [inline source/backend/api/Areas/Projects/Controllers/ProjectController.cs:30]: I am not a fan of the pragmas/regions but they are standard in our code.

File: source/backend/api/Areas/Projects/Models/ProjectSearchModel.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
+using System;
+
+namespace Pims.Api.Areas.Projects.Models
+{
+    public class ProjectSearchModel
```
Comment (FuriousLlama) [inline source/backend/api/Areas/Projects/Models/ProjectSearchModel.cs:5]: This model looks pretty similar to the existing ProjectModel. It would be better if we have only one given the data is pretty much the same

File: source/backend/api/Models/Concepts/Project/ProjectModel.cs [hunk 3]
```diff
@@ -48,6 +50,16 @@ public class ProjectModel : BaseModel
...
+
+        /// <summary>
+        /// get/set - Last Updated Date.
+        /// </summary>
+        public DateTime AppLastUpdateTimestamp { get; set; }
+
+        /// <summary>
+        /// get/set - Last Updated by.
+        /// </summary>
+        public string AppLastUpdateUserid { get; set; }
```
Comment (FuriousLlama) [inline source/backend/api/Models/Concepts/Project/ProjectModel.cs:62]: These fields should come in the base Model already

File: source/backend/api/Models/RegionModel.cs [hunk 4]
```diff
@@ -0,0 +1,18 @@
+namespace Pims.Api.Models
+{
+    /// <summary>
+    /// Provides a lease-oriented insurance model.
+    /// </summary>
+    public class RegionModel : BaseModel
```
Comment (FuriousLlama) [inline source/backend/api/Models/RegionModel.cs:6]: The address uses CodeTypeModel. Would that model suffice?

File: source/backend/tests/unit/api/Services/ProjectServiceTest.cs [hunk 5]
```diff
@@ -66,6 +65,83 @@ public void Search_NoPermission()
...
+            var repository = helper.GetService<Mock<IProjectRepository>>();
+            repository.Setup(x => x.GetPageAsync(It.IsAny<ProjectFilter>()))
+                .ReturnsAsync(new Paged<PimsProject>()
+                {
+                    Page = 1,
+                });
+
+
+            // Act
+            var sut = await service.GetPage(new ProjectFilter { ProjectName = "test" });
```
Comment (FuriousLlama) [inline source/backend/tests/unit/api/Services/ProjectServiceTest.cs:140]: nit: sut -> result

File: source/frontend/src/AppRouter.tsx [hunk 6]
```diff
@@ -239,6 +240,14 @@ const AppRouter: React.FC<React.PropsWithChildren<unknown>> = () => {
             claim={Claims.ACQUISITION_VIEW}
             title={getTitle('View Acquisition Files')}
           />
+          <AppRoute
+            protected
+            path="/project/list"
+            customComponent={ProjectListView}
+            layout={AuthLayout}
+            //claim={Claims.PROJECT_VIEW}
```
Comment (FuriousLlama) [inline source/frontend/src/AppRouter.tsx:248]: claim missing

File: source/frontend/src/features/projects/list/ProjectListView.scss [hunk 7]
```diff
@@ -0,0 +1,4 @@
+@import '~assets/scss/variables';
+@import '~assets/scss/colors';
+
+$contentheading-height: 6rem;
```
Comment (FuriousLlama) [inline source/frontend/src/features/projects/list/ProjectListView.scss:4]: Not sure this file is needed given most of the styles we use are directly on the page itself

File: source/frontend/src/features/projects/styles.ts [hunk 8]
```diff
@@ -0,0 +1,51 @@
+import { Button } from 'components/common/buttons/Button';
+import { InlineForm, InlineInput } from 'components/common/form/styles';
+import { Scrollable as ScrollableBase } from 'components/common/Scrollable/Scrollable';
+import styled from 'styled-components';
+
+export const ListPage = styled.div`
```
Comment (FuriousLlama) [inline source/frontend/src/features/projects/styles.ts:6]: can the styles in this page be reused from the other search/list components?

File: source/frontend/src/interfaces/IProjectSearchResults.ts [hunk 9]
```diff
@@ -0,0 +1,9 @@
+export interface IProjectSearchResult {
```
Comment (FuriousLlama) [inline source/frontend/src/interfaces/IProjectSearchResults.ts:1]: same as the backend. This could simply be the projectModel. Also, models from the api start with "API_*"

File: source/frontend/src/models/api/Project.ts [hunk 10]
```diff
@@ -8,10 +8,12 @@ export interface Api_Project extends Api_ConcurrentVersion {
   businessFunctionCode?: any; // TODO: Match new code types
   costTypeCode?: any; // TODO: Match new code types
   workActivityCode?: any; // TODO: Match new code types
-  regionCode?: number;
+  regionCode?: Api_TypeCode<number>;
```
Comment (FuriousLlama) [inline source/frontend/src/models/api/Project.ts:11]: This type does not match the backend.

File: source/backend/api/Mapping/RegionMap.cs [hunk 11]
```diff
@@ -0,0 +1,16 @@
+using Mapster;
+using Pims.Api.Models;
+using Entity = Pims.Dal.Entities;
+
+namespace Pims.Api.Mapping
+{
+    public class RegionMap : IRegister
```
Comment (devinleighsmith) [inline source/backend/api/Mapping/RegionMap.cs:7]: Region inherits from ICodeEntity so you may not need this. See AddressMap or  LookupMap

File: source/backend/api/Models/RegionModel.cs (hunk 4 above)
Comment (devinleighsmith) [reply source/backend/api/Models/RegionModel.cs:6]: yes, either codeTypeModal or just TypeModel

File: source/backend/dal/Repositories/ProjectRepository.cs [hunk 12]
```diff
@@ -43,6 +44,79 @@ public IList<PimsProject> SearchProjects(string filter, int maxResult)
...
+            {
+                query = query.Where(x => x.ProjectStatusTypeCodeNavigation.ProjectStatusTypeCode == filter.ProjectStatusCode);
+            }
+
+            if (!string.IsNullOrWhiteSpace(filter.ProjectRegionCode))
+            {
+                query = query.Where(x => x.RegionCode.ToString() == filter.ProjectRegionCode);
+            }
+
+            if (filter.Sort?.Any() == true)
```
Comment (devinleighsmith) [inline source/backend/dal/Repositories/ProjectRepository.cs:96]: this and the next section appear duplicated.

File: source/backend/entities/Partials/ProjectStatusType.cs [hunk 13]
```diff
@@ -27,6 +22,15 @@ public PimsProjectStatusType(string id)
         {
             Id = id;
         }
-        #endregion
+
+        public static class PimsProjectStatusTypes
```
Comment (devinleighsmith) [inline source/backend/entities/Partials/ProjectStatusType.cs:26]: I don't see this being used anywhere, why is this here?

File: source/backend/tests/unit/api/Services/ProjectServiceTest.cs [hunk 14]
```diff
@@ -66,6 +65,83 @@ public void Search_NoPermission()
...
+            // Arrange
+            var helper = new TestHelper();
+            var user = PrincipalHelper.CreateForPermission(Permissions.ProjectView);
+            var service = helper.Create<ProjectService>(user);
+
+            var repository = helper.GetService<Mock<IProjectRepository>>();
+            repository.Setup(x => x.GetPageAsync(It.IsAny<ProjectFilter>()))
+                .ReturnsAsync(new Paged<PimsProject>()
+                {
+                    Page = 1,
```
Comment (devinleighsmith) [inline source/backend/tests/unit/api/Services/ProjectServiceTest.cs:135]: is the page necessary?

File: source/frontend/src/features/projects/list/ProjectFilter/ProjectFilter.tsx [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+  React.PropsWithChildren<IProjectFilterProps>
+> = ({ filter, setFilter }) => {
+  const [selectedStatus, setInitialSelectedStatus] = useState<string>(ProjectStatusTypes.Active);
+
+  useEffect(() => {
+    setInitialSelectedStatus(ProjectStatusTypes.Active);
+  }, []);
+
+  const onSearchSubmit = (values: IProjectFilter, { setSubmitting }: any) => {
+    console.log(values);
```
Comment (devinleighsmith) [inline source/frontend/src/features/projects/list/ProjectFilter/ProjectFilter.tsx:40]: please remove console logs used for debugging when submitting PRs.

File: source/frontend/src/features/projects/list/ProjectFilter/ProjectFilter.tsx [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+ * Filter bar for Projects.
+ * @param {IProjectFilterProps} props
+ */
+export const ProjectFilter: React.FunctionComponent<
+  React.PropsWithChildren<IProjectFilterProps>
+> = ({ filter, setFilter }) => {
+  const [selectedStatus, setInitialSelectedStatus] = useState<string>(ProjectStatusTypes.Active);
+
+  useEffect(() => {
+    setInitialSelectedStatus(ProjectStatusTypes.Active);
```
Comment (devinleighsmith) [inline source/frontend/src/features/projects/list/ProjectFilter/ProjectFilter.tsx:36]: this seems redundant, as the useState is initialized to the value of ProjectStatusTypes.Active but you are setting it to active again here once during component initialization.

File: source/frontend/src/features/projects/list/ProjectFilter/ProjectFilter.tsx [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+};
+
+/**
+ * Filter bar for Projects.
+ * @param {IProjectFilterProps} props
+ */
+export const ProjectFilter: React.FunctionComponent<
+  React.PropsWithChildren<IProjectFilterProps>
+> = ({ filter, setFilter }) => {
+  const [selectedStatus, setInitialSelectedStatus] = useState<string>(ProjectStatusTypes.Active);
```
Comment (devinleighsmith) [inline source/frontend/src/features/projects/list/ProjectFilter/ProjectFilter.tsx:33]: I don't think using state here makes sense. If you want the default value to be active I would recommend setting that in whatever parent is passing the filter prop here.

File: source/frontend/src/features/projects/list/ProjectListView.scss (hunk 7 above)
Comment (devinleighsmith) [reply source/frontend/src/features/projects/list/ProjectListView.scss:4]: Yes, also in general we use styled components not scss for styling.

File: source/frontend/src/features/projects/list/ProjectListView.tsx [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+    (filter: IProjectFilter) => {
+      setFilter(filter);
+      setCurrentPage(0);
+    },
+    [setFilter, setCurrentPage],
+  );
+
+  useEffect(() => {
+    if (error) {
+      toast.error(error?.message);
```
Comment (devinleighsmith) [inline source/frontend/src/features/projects/list/ProjectListView.tsx:52]: is this part of a pattern we are using elsewhere? seems odd.

File: source/backend/api/Models/Concepts/Project/ProjectModel.cs (hunk 3 above)
Comment (eddherrera) [reply source/backend/api/Models/Concepts/Project/ProjectModel.cs:62]: It didn't. Replaced with 'BaseAppModel' which does.

File: source/frontend/src/interfaces/IProjectSearchResults.ts (hunk 9 above)
Comment (devinleighsmith) [reply source/frontend/src/interfaces/IProjectSearchResults.ts:1]: Yes interfaces starting with I are frontend use only interfaces (not api related)

File: source/frontend/src/models/api/Project.ts (hunk 10 above)
Comment (devinleighsmith) [reply source/frontend/src/models/api/Project.ts:11]: but you could update the backend to match.

**Review by devinleighsmith — CHANGES_REQUESTED** [review]
see comments

File: source/frontend/src/features/projects/list/ProjectListView.tsx (hunk 18 above)
Comment (eddherrera) [reply source/frontend/src/features/projects/list/ProjectListView.tsx:52]: LeaseListView uses it. Will remove it for now.

File: source/backend/api/Areas/Projects/Models/ProjectSearchModel.cs (hunk 2 above)
Comment (eddherrera) [reply source/backend/api/Areas/Projects/Models/ProjectSearchModel.cs]: Deleted model

File: source/backend/api/Models/Concepts/Project/ProjectModel.cs (hunk 3 above)
Comment (eddherrera) [reply source/backend/api/Models/Concepts/Project/ProjectModel.cs]: Updated

File: source/backend/api/Models/RegionModel.cs (hunk 4 above)
Comment (eddherrera) [reply source/backend/api/Models/RegionModel.cs:6]: Updated

File: source/backend/tests/unit/api/Services/ProjectServiceTest.cs (hunk 5 above)
Comment (eddherrera) [reply source/backend/tests/unit/api/Services/ProjectServiceTest.cs]: updated

File: source/frontend/src/AppRouter.tsx (hunk 6 above)
Comment (eddherrera) [reply source/frontend/src/AppRouter.tsx]: Updated

File: source/frontend/src/features/projects/list/ProjectSearchResults/ProjectSearchResults.tsx [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+  pageIndex?: number;
+  sort?: TableSort<ProjectSearchResultModel>;
+  setSort: (value: TableSort<ProjectSearchResultModel>) => void;
+  setPageSize?: (value: number) => void;
+  setPageIndex?: (value: number) => void;
+  loading?: boolean;
+}
+
+/**
+ * Component that renders search results for acquisition files.
```
[the remark this answers is not in the document]
Comment (eddherrera) [reply source/frontend/src/features/projects/list/ProjectSearchResults/ProjectSearchResults.tsx]: updated

File: source/frontend/src/models/api/Project.ts (hunk 10 above)
Comment (eddherrera) [reply source/frontend/src/models/api/Project.ts:11]: Updated

File: source/backend/api/Mapping/RegionMap.cs (hunk 11 above)
Comment (eddherrera) [reply source/backend/api/Mapping/RegionMap.cs:7]: Updated

File: source/backend/api/Models/Concepts/Project/ProjectModel.cs [hunk 20]
```diff
@@ -32,7 +34,7 @@ public class ProjectModel : BaseModel
         /// <summary>
         /// get/set - The region code.
         /// </summary>
-        public short RegionCode { get; set; } // TODO: Use the correct type.
+        public RegionModel RegionCode { get; set; } // TODO: Use the correct type.
```
[the remark this answers is not in the document]
Comment (eddherrera) [reply source/backend/api/Models/Concepts/Project/ProjectModel.cs]: Updated

File: source/backend/entities/Partials/ProjectStatusType.cs (hunk 13 above)
Comment (eddherrera) [reply source/backend/entities/Partials/ProjectStatusType.cs]: removed

File: source/backend/tests/unit/api/Services/ProjectServiceTest.cs (hunk 14 above)
Comment (eddherrera) [reply source/backend/tests/unit/api/Services/ProjectServiceTest.cs:135]: Yes, it would throw an excpetion otherwise.

File: source/frontend/src/features/projects/list/ProjectFilter/ProjectFilter.tsx (hunk 17 above)
Comment (eddherrera) [reply source/frontend/src/features/projects/list/ProjectFilter/ProjectFilter.tsx]: passed with prop

File: source/frontend/src/features/projects/styles.ts (hunk 8 above)
Comment (eddherrera) [reply source/frontend/src/features/projects/styles.ts:6]: Will address on next PR.

**Review by devinleighsmith — APPROVED** [review]
